### PR TITLE
NIFI-1537 Added SNMP processors

### PR DESCRIPTION
**Pull request by pvillard31**
Processors for SNMP set/get/walk operations over SNMP v1, v2c & v3 using SNMP4J library.

### CODE REVIEW

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPUtils.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+     * @param processSession instance of {@link ProcessSession}
+     * @return updated {@link FlowFile}
+     */
+    public static FlowFile updateFlowFileAttributesWithPduProperties(PDU response, FlowFile flowFile, ProcessSession processSession) {
+        if (response != null) {
+            try {
+                Method[] methods = PDU.class.getDeclaredMethods();
+                Map<String, String> attributes = new HashMap<String, String>();
+                for (Method method : methods) {
+                    if (Modifier.isPublic(method.getModifiers()) && (method.getParameterCount() == 0) && method.getName().startsWith("get")) {
```
Comment (olegz) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPUtils.java]: Pierre, I noticed this when I was researching for your OnStopped issue. I believe this method 'getParameterCount' was introduced in Java 8 while we still have to be compliant with Java 7. So consider refactoring.

Comment (pvillard31) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPUtils.java]: Thanks Oleg, I have fixed it.

**Comment by olegz** [comment]
Pierre, so the ReflectionUtils fix is in https://github.com/apache/nifi/pull/260. So watch the PR and once it is merged, rebase, update (remove that temp fix code) and test and let me know. I've already tested it with your branch but still extra validation would help.

**Comment by pvillard31** [comment]
Oleg, yes sure, will do!

**Comment by pvillard31** [comment]
Pre-requisite: NIFI-1595 has to be merged in before

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+     */
+    private void buildTargetResource(ProcessContext context) {
+        if((this.transportMapping == null) || !this.transportMapping.isListening() || (this.snmp == null)) {
+            try {
+                this.transportMapping = new DefaultUdpTransportMapping();
+                this.snmp = new Snmp(this.transportMapping);
+
+                if("SNMPv3".equals(context.getProperty(SNMP_VERSION).getValue())) {
+                    USM usm = new USM(SecurityProtocols.getInstance(), new OctetString(MPv3.createLocalEngineID()), 0);
+                    SecurityModels.getInstance().addSecurityModel(usm);
```
Comment (olegz) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java]: I just initiated a DISCUSS thread on the dev list. Basically while there is not much we can do at the moment, the above code will set the security model that will be applicable across all instances of SNMP processors within a single instance of NiFi.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/java/org/apache/nifi/snmp/processors/SetSNMPTest.java [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+    //            runner.setProperty(SetSNMP.SNMP_VERSION, "SNMPv1");
+    //
+    //            runner.run();
+    //            Thread.sleep(200);
+    //            final MockFlowFile successFF = runner.getFlowFilesForRelationship(SetSNMP.REL_SUCCESS).get(0);
+    //            assertNotNull(successFF);
+    //            assertEquals("test", successFF.getAttributes().get(SNMPUtils.SNMP_PROP_PREFIX + sysDescr.toString()));
+    //        }
+    //    }
+
```
Comment (olegz) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/java/org/apache/nifi/snmp/processors/SetSNMPTest.java]: Pierre, did you intent to keep commented code above?

Comment (pvillard31) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/java/org/apache/nifi/snmp/processors/SetSNMPTest.java]: Definitely not, my bad. Going to fix it.

**Comment by olegz** [comment]
Pierre

You may want to rebase (merge conflicts) and push again since it's out of sink.
Also, aside from unit tests would you mind giving some pointers on how to test it outside of unit tests? A link on how to setup an environment or a short writeup. I am not that familiar with SNMP

**Comment by michalklempa** [comment]
@olegz there is http://snmpsim.sourceforge.net/ project, which can be used for testing purposes. Some simulation files are available with distribution, so basic testing may be done out-of-the-box

**Comment by pvillard31** [comment]
@olegz I just rebased the PR.

As @michalklempa mentioned, you can use online SNMP service provided by snmpsim (http://snmpsim.sourceforge.net/public-snmp-simulator.html), otherwise it depends of your OS. Personally, I tried the processor by requesting basic SNMP data (such as uptime for example) on a CentOS machine with net-snmp agent running with various configurations in SNMP v1, v2c and v3 (MD5/DES). 

As a side note, at this moment, the processor is expecting an OID in numerical format. If I try the following command on my computer:

`snmpwalk -v2c -c public demo.snmplabs.com IF-MIB::interfaces`

I can get numerical OID by adding the option -On:

`snmpwalk -v2c -c public demo.snmplabs.com IF-MIB::interfaces -On`

This way I can see that IF-MIB::interfaces is equivalent to .1.3.6.1.2.1.2.2.1

I will try to write something to help setting an environment. On a Linux-like environment, a quick setup can be:
yum install net-snmp net-snmp-utils
service snmpd start

It will allow access to basic SNMP information in v1 and v2c with 'public' as community. For SNMP v3 setup, it is possible to add users using the command net-snmp-create-v3-user.

Hope it helps, if not, let me know ;)

**Comment by olegz** [comment]
@michalklempa @pvillard31 thank you both for the input. Will be reviewing and testing.

**Comment by trixpan** [comment]
@olegz  any luck with this review? Cheers

**Comment by pvillard31** [comment]
@trixpan @olegz @michalklempa 
I rebased the PR against master

**Comment by olegz** [comment]
@pvillard31 could you please rebase and fix merge conflicts one last time. I am now committed to review/test it so we can have it in 0.7.

**Comment by pvillard31** [comment]
@olegz half-time ;) done!

**Comment by olegz** [comment]
. . . and you almost scored ;) but. . . anyway, will try to wrap it up tomorrow

**Comment by olegz** [comment]
Ok, finally I am in this PR and no distractions ;)

**Comment by olegz** [comment]
@pvillard31  Had a merge conflict again, but fixed it myself. Nothing to do, just making comment for traceability.

**Pull request closed without merging**